### ISSUE TITLE
[Feature Request]: Fusion search engine for monolith rspress projects

### ISSUE THREAD

**Issue by zoolsher** [web-infra-dev/rspress#1297]
### What problem does this feature solve?

In complex and large-scale projects, the current search functionality lacks classification capabilities, making it difficult for users to efficiently find the information they need. As the volume of project data increases, unclassified search results can get lost in the vast sea of information, leading to imprecise search results and decreased user productivity. By providing a search feature with classification capabilities, users can more accurately and quickly find the information they need based on specific categories. This makes the search process more purposeful and efficient, thereby improving user experience and work efficiency.

#### add category to meta

step 1. Configure all available categories in `rspress.config.js`

```ts
export default {
  search: {
    categories: ['css', 'html', 'foo', 'bar']
  }
}
```

step 2. Declare a document’s category

```markdown
---
category: ['css', 'html']
category: 'css'
---

The markdown content
```

#### UI for search with category
![Search With Category](https://github.com/user-attachments/assets/41da1d32-f3c2-4cea-bec3-0646f8f58495)

#### UI for search without category
![Search Without Category](https://github.com/user-attachments/assets/a29e265a-9e05-4dd3-a2fa-44b37d588b2f)

#### Can document writers add any category they wish?

No.

The categories are defined by the search engine, so they must be a fixed set controlled by the project configuration.

#### What if a markdown file does not specify a category?

It will automatically **extend** from its parent folder. The root folder will have a default category, which is the first one in the categories set by default. You can also specify this with the following configuration:

```ts
export default {
  search: {
    categories: ['css', 'html', 'foo', 'bar'],
    defaultCategory: 'foo', // defaults to 'css'
  }
}
```

#### Support for custom sources?

If a developer wants to add custom categories such as `ai`, `discord qa channel`, or `github issues`, where the name, content, and display are all customized, they can use the category plugin feature. Here’s how:
```ts
export default {
  search: {
    categories: ['css', 'html', 'foo', 'bar'],
    categoryPlugins: [{ 
      name: "baz",
      search: (context) => {
        // Code to execute when a user searches within this category
        return SomeReactComponent;
      }
    }],
  }
}
```

**Comment by chenjiahan** [comment]
Declaring a category for each document is quite time-consuming. We should allow declaring categories for all documents in a specified directory.

This can be achieved through Rspress's `meta.json` file:

```jsonc
[
  {
    "type": "dir",
    "name": "css",
    "label": "CSS",
    // I prefer using `searchCategory` instead of `category`, because it is more explicit
    "searchCategory": "css"
  }
]
```

> https://rspress.dev/guide/basic/auto-nav-sidebar

**Comment by zoolsher** [comment]
> Declaring a category for each document is quite time-consuming. We should allow declaring categories for all documents in a specified directory.
> 
> This can be achieved through Rspress's `meta.json` file:
> 
> ```js
> [
>   {
>     "type": "dir",
>     "name": "css",
>     "label": "CSS",
>     // I prefer using `searchCategory` instead of `category`, because it is more explicit
>     "searchCategory": "css"
>   }
> ]
> ```
> 
> > https://rspress.dev/guide/basic/auto-nav-sidebar

Yes, I have included it in the QA section. It will automatically extend from the parent folder.

**Comment by chenjiahan** [comment]
For the `categoryPlugins`, we can passing the path of the React component instead of importing thg React component. Because `rspress.config.ts` is a Node.js file and should avoid importing React components.

```diff
export default {
  search: {
    categories: ['css', 'html', 'foo', 'bar'],
    categoryPlugins: [{ 
      name: "baz",
-      search: (context) => {
-        // Code to execute when a user searches within this category
-        return SomeReactComponent;
-      }
+     search: path.join(__dirname, '/path/to/search.tsx'),
    }],
  }
}
```

**Comment by zoolsher** [comment]
> For the `categoryPlugins`, we can passing the path of the React component instead of importing thg React component. Because `rspress.config.ts` is a Node.js file and should avoid importing React components.
> 
> ```diff
> export default {
>   search: {
>     categories: ['css', 'html', 'foo', 'bar'],
>     categoryPlugins: [{ 
>       name: "baz",
> -      search: (context) => {
> -        // Code to execute when a user searches within this category
> -        return SomeReactComponent;
> -      }
> +     search: path.join(__dirname, '/path/to/search.tsx'),
>     }],
>   }
> }
> ```

Yes. I'm just trying to make it easier to understand what happened in another file

**Comment by chenjiahan** [comment]
Nice👍  Let's do it

**Comment by Timeless0911** [comment]
If a mdx file in a folder needs to override some categories already set in folder dimension, we can use both frontmatter or `_meta.json` to config?

**Comment by zoolsher** [comment]
I believe we can keep it simple like this.
```ts
// fake code
thisDocCategory = mdFileHeader.category ?? parentFolderCategory
```

**Comment by Timeless0911** [comment]
> I believe we can keep it simple like this.
> 
> ```ts
> // fake code
> thisDocCategory = mdFileHeader.category ?? parentFolderCategory
> ```

Where does `parentFolderCategory` come from? You mean every folder should have a same name mdx file to config `category` for the whole folder, or all files in this folder should use the default value of `search.defaultCategory`? Maybe some folder do not have a index entry mdx file.

**Comment by SoonIter** [comment]
can be solved via search hooks customization

- https://rspress.rs/guide/advanced/custom-search#understanding-searchhooks

<img width="510" height="535" alt="Image" src="https://github.com/user-attachments/assets/1bf1b3c4-ab78-4658-bf41-c828dd5e58a9" />
